osf/utils/registrations.py

```diff
@@ -0,0 +1,174 @@
+    # stored under extra
+    if block_type == 'file-input' and not keys:
+        extra = format_extra(keyed_value.get('extra', []))
+        return extra
+    else:
+        value = keyed_value.get('value', '')
+        return value
+
+# For flatten_registration_metadata
+def get_nested_answer(nested_response, block_type, keys):
+    """
+    Recursively fetches the nested response in registered_meta.
+
+    :params nested_response dictionary
+    :params keys array, of nested question_ids: ["recommended-analysis", "specify", "question11c"]
+    :returns array (files or multi-response answers) or a string
+    """
+    if isinstance(nested_response, dict):
+        key = keys.pop(0)
+        # Returns the value associated with the given key
+        value = get_value_or_extra(nested_response, block_type, key, keys)
+        return get_nested_answer(value, block_type, keys)
+    else:
+        # Once we've drilled down through the entire dictionary, our nested_response
+        # should be an array or a string
+        return nested_response
+
+# For expanding registration_responses
+def set_nested_values(nested_dictionary, keys, value):
+    """
+    Drills down through the nested dictionary, accessing each key in the array,
+    and sets the last key equal to the  passed in value, if this key doesn't already exist.
+
+    Assumes all keys are present, except for potentially the final key.
+
+    :param nested_dictionary, dictionary
+    :param keys, array
+    :param value, object, array, or string
+    """
+    for key in keys[:-1]:
+        nested_dictionary = nested_dictionary.get(key, None)
+
+    final_key = keys[-1]
+    if not nested_dictionary.get(final_key):
+        nested_dictionary[final_key] = value
+
+# For expanding registration_responses
+def build_answer_block(block_type, value):
+    extra = []
+    if block_type == 'file-input':
+        extra = value
+        value = ''
+        for file in extra:
+            if file.get('file_name'):
+                file['data'] = {
+                    'name': file.get('file_name')  # What legacy FE needs for rendering
+                }
+    return {
+        'comments': [],
+        'value': value,
+        'extra': extra
+    }
+
+# For expanding registration_responses
+def build_registration_metadata_dict(keys, current_index=0, metadata={}, value={}):
+    """
+    Function will recursively loop through each key in the list, checking if it exists in metadata, if not,
+    adding another nested level in the dictionary.
+
+    For example, calling build_registration_metadata_dict(
+        ["recommended-analysis", "value", "specify", "value", "question11c"], value='hello'), yields,
+
+    {
+        'recommended-analysis': {
+            'value': {
+                'specify': {
+                    'value': {
+                        'question11c': 'hello'
+                    }
+                }
+            }
+        }
+    }
```

Comment:
This also assumes that the deepest level of nesting, contains data at either `value` or `extra`, not both.  So if there is files information under `extra`, it is not storing information under `value`.  Double check prod data to make sure this isn't an issue.

Reply:
@aaxelb - just pointing this out, for prereg challenge people might be able to add a value to a file, so there might be data in both places. It might not be worth preserving this information, but let's check.  My current implementation does *not* assume that data is in both places.